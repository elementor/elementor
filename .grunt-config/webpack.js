/**
 * Grunt webpack task config
 * @package Elementor
 */
const path = require( 'path' );

const packagesConfigs = require( './webpack.packages.js' );

// Handling minification for production assets.
const TerserPlugin = require( 'terser-webpack-plugin' );

const aliasList = require('./webpack.alias.js').resolve;

const webpack = require('webpack');

// Cleaning up existing chunks before creating new ones.
const RemoveChunksPlugin = require('./remove-chunks');

const WatchTimePlugin = require('./plugins/watch-time/index');

// Preventing auto-generated long names of shared sub chunks (optimization.splitChunks.minChunks) by using only the hash.
const getChunkName = ( chunkData, environment ) => {
	const minSuffix = 'production' === environment ? '.min' : '',
		name = chunkData.chunk.name ? '[name].' : '';

	return `${ name }[contenthash].bundle${ minSuffix }.js`;
};

const getModuleRules = ( presets ) => {
	return {
		rules: [
			{
				test: /app.*\.(s)?css$/i,
				use: [
					{
						loader: './loaders/app-imports.js',
					},
				],
			},
			{
				test: /\.js$/,
				exclude: /node_modules/,
				use: [
					{
						loader: 'babel-loader',
						options: {
							presets,
							plugins: [
								[ '@wordpress/babel-plugin-import-jsx-pragma' ],
								[ '@babel/plugin-transform-react-jsx', {
									'pragmaFrag': 'React.Fragment',
								} ],
								[ '@babel/plugin-transform-runtime' ],
								[ '@babel/plugin-transform-modules-commonjs' ],
							],
						},
					},
				],
			},
		],
	};
};

const moduleRules = getModuleRules( [ '@wordpress/default' ] );

const frontendRulesPresets = [ [
	'@babel/preset-env',
	{
		targets: {
			browsers: [
				'last 1 Android versions',
				'last 1 ChromeAndroid versions',
				'last 2 Chrome versions',
				'last 2 Firefox versions',
				'Safari >= 14',
				'iOS >= 14',
				'last 2 Edge versions',
				'last 2 Opera versions',
			],
		},
		"useBuiltIns": "usage",
		"corejs": "3.23",
	}
] ];

const frontendModuleRules = getModuleRules( frontendRulesPresets );

const entry = {
	'editor': [
		path.resolve( __dirname, '../assets/dev/js/editor/utils/jquery-serialize-object.js' ),
		path.resolve( __dirname, '../assets/dev/js/editor/utils/jquery-html5-dnd.js' ),
		path.resolve( __dirname, '../assets/dev/js/editor/editor.js' ),
	],
	'admin': path.resolve( __dirname, '../assets/dev/js/admin/admin.js' ),
	'admin-feedback': path.resolve( __dirname, '../assets/dev/js/admin/admin-feedback.js' ),
	'announcements-app': path.resolve( __dirname, '../modules/announcements/assets/js/index.js' ),
	'common': path.resolve( __dirname, '../core/common/assets/js/common.js' ),
	'dev-tools': path.resolve( __dirname, '../modules/dev-tools/assets/js/index.js' ),
	'elementor-admin-bar': path.resolve( __dirname, '../modules/admin-bar/assets/js/frontend/module.js' ),
	'gutenberg': path.resolve( __dirname, '../assets/dev/js/admin/gutenberg.js' ),
	'new-template': path.resolve( __dirname, '../assets/dev/js/admin/new-template/new-template.js' ),
	'app': path.resolve( __dirname, '../app/assets/js/index.js' ),
	'app-loader': path.resolve( __dirname, '../app/assets/js/app-loader' ),
	'app-packages': path.resolve( __dirname, '../app/assets/js/app-packages' ),
	'beta-tester': path.resolve( __dirname, '../assets/dev/js/admin/beta-tester/beta-tester.js' ),
	'common-modules': path.resolve( __dirname, '../core/common/assets/js/modules' ),
	'editor-modules': path.resolve( __dirname, '../assets/dev/js/editor/modules.js' ),
	'admin-modules': path.resolve( __dirname, '../assets/dev/js/admin/modules.js' ),
	'editor-document': path.resolve( __dirname, '../assets/dev/js/editor/editor-document.js' ),
	'qunit-tests': path.resolve( __dirname, '../tests/qunit/main.js' ),
	'admin-top-bar': path.resolve( __dirname, '../modules/admin-top-bar/assets/js/admin.js' ),
	'checklist': path.resolve( __dirname, '../modules/checklist/assets/js/editor.js' ),
	'nested-elements': path.resolve( __dirname, '../modules/nested-elements/assets/js/editor/index.js' ),
	'nested-tabs': path.resolve( __dirname, '../modules/nested-tabs/assets/js/editor/index.js' ),
	'nested-accordion': path.resolve( __dirname, '../modules/nested-accordion/assets/js/editor/index.js' ),
	'container-converter': path.resolve( __dirname, '../modules/container-converter/assets/js/editor/module.js' ),
	'atomic-widgets-editor': path.resolve( __dirname, '../modules/atomic-widgets/assets/js/editor/module.js' ),
	'notes': path.resolve( __dirname, '../modules/notes/assets/js/notes.js' ),
	'web-cli': path.resolve( __dirname, '../modules/web-cli/assets/js/index.js' ),
	'import-export-admin': path.resolve( __dirname, '../app/modules/import-export/assets/js/admin.js' ),
	'kit-elements-defaults-editor': path.resolve( __dirname, '../modules/kit-elements-defaults/assets/js/editor/index.js' ),
	'editor-loader-v1': path.resolve( __dirname, '../core/editor/loader/v1/js/editor-loader-v1.js' ),
	'editor-loader-v2': path.resolve( __dirname, '../core/editor/loader/v2/js/editor-loader-v2.js' ),
	'editor-environment-v2': path.resolve( __dirname, '../core/editor/loader/v2/js/editor-environment-v2.js' ),
	'responsive-bar': path.resolve( __dirname, '../assets/dev/js/editor/regions/responsive-bar/index.js' ),
	'ai': path.resolve( __dirname, '../modules/ai/assets/js/editor/index.js' ),
	'admin-notifications': path.resolve( __dirname, '../modules/notifications/assets/js/admin.js' ),
	'editor-notifications': path.resolve( __dirname, '../modules/notifications/assets/js/editor.js' ),
	'ai-layout': path.resolve( __dirname, '../modules/ai/assets/js/editor/layout-module.js' ),
	'ai-gutenberg': path.resolve( __dirname, '../modules/ai/assets/js/gutenberg/index.js' ),
	'element-manager-admin': path.resolve( __dirname, '../modules/element-manager/assets/js/admin.js' ),
	'media-hints': path.resolve( __dirname, '../assets/dev/js/admin/hints/media.js' ),
	'ai-media-library': path.resolve( __dirname, '../modules/ai/assets/js/media-library/index.js' ),
	// Temporary solution for the AI App in the Admin.
	'ai-admin': path.resolve( __dirname, '../modules/ai/assets/js/admin/index.js' ),
	'styleguide': path.resolve( __dirname, '../modules/styleguide/assets/js/styleguide.js' ),
	'styleguide-app-initiator': path.resolve( __dirname, '../modules/styleguide/assets/js/styleguide-app-initiator.js' ),
	'e-home-screen': path.resolve( __dirname, '../modules/home/assets/js/app.js' ),
	'e-wc-product-editor': path.resolve( __dirname, '../modules/wc-product-editor/assets/js/e-wc-product-editor.js' ),
	'floating-elements-modal': path.resolve( __dirname, '../assets/dev/js/admin/floating-elements/new-floating-elements.js' ),
};

const frontendEntries = {
	'frontend-modules': path.resolve( __dirname, '../assets/dev/js/frontend/modules.js' ),
	'frontend': { import: path.resolve( __dirname, '../assets/dev/js/frontend/frontend.js' ), dependOn: 'frontend-modules' },
};

const externals = [
	{
		'@wordpress/i18n': 'wp.i18n',
		react: 'React',
		'react-dom': 'ReactDOM',
		'@elementor/app-ui': 'elementorAppPackages.appUi',
		'@elementor/components': 'elementorAppPackages.components',
		'@elementor/hooks': 'elementorAppPackages.hooks',
		'@elementor/site-editor': 'elementorAppPackages.siteEditor',
		'@elementor/router': 'elementorAppPackages.router',
		'@elementor/ui': 'elementorV2.ui',
		'@elementor/icons': 'elementorV2.icons',
		'@elementor/editor-app-bar': 'elementorV2.editorAppBar',
		'@elementor/editor-v1-adapters': 'elementorV2.editorV1Adapters',
		'@wordpress/dom-ready': 'wp.domReady',
		'@wordpress/components': 'wp.components',
		'@wordpress/core-data': 'wp.coreData',
		'@wordpress/data': 'wp.data',
		'@wordpress/plugins': 'wp.plugins',
		'@woocommerce/admin-layout': 'wc.adminLayout',
	},
	// Handle tree-shaking imports for ui and icons packages (@elementor/ui/xxx) to be pointed to the external object (elementorV2.ui.xxx).
	function ( { request }, callback ) {
		const matches = request.match( /^@elementor\/(ui|icons)\/(.+)$/ );

		if ( matches?.length ) {
			return callback( null, `elementorV2.${ matches[ 1 ] }['${ matches[ 2 ] }']` );
		}

		callback();
	},
];

const plugins = [
	new webpack.ProvidePlugin( {
		React: 'react',
		ReactDOM: 'react-dom',
		PropTypes: 'prop-types',
		__: ['@wordpress/i18n', '__'],
		sprintf: ['@wordpress/i18n', 'sprintf'],
	} ),
	new WatchTimePlugin(),
];

const baseConfig = {
	target: 'web',
	context: __dirname,
	externals,
	resolve: aliasList,
};

const devSharedConfig = {
	...baseConfig,
	devtool: 'source-map',
	mode: 'development',
	output: {
		path: path.resolve( __dirname, '../assets/js' ),
		chunkFilename: ( chunkData ) => getChunkName( chunkData, 'development' ),
		filename: '[name].js',
		devtoolModuleFilenameTemplate: '../[resource]',
		// Prevents the collision of chunk names between different bundles.
		uniqueName: 'elementor',
	},
	watch: true,
};

const webpackConfig = [
	{
		...devSharedConfig,
		module: moduleRules,
		plugins: [
			...plugins,
		],
		name: 'base',
		entry: entry,
	},
	{
		...devSharedConfig,
		module: frontendModuleRules,
		plugins: [
			new RemoveChunksPlugin( '.bundle.js' ),
			...plugins,
		],
		name: 'frontend',
		optimization: {
			runtimeChunk:  {
				name: 'webpack.runtime',
			},
			splitChunks: {
				minChunks: 2,
			},
		},
		entry: frontendEntries,
	},
	packagesConfigs.dev,
];

const prodSharedOptimization = {
	minimize: true,
	minimizer: [
		new TerserPlugin( {
			terserOptions: {
				keep_fnames: true,
			},
			include: /\.min\.js$/
		} ),
	],
};

const prodSharedConfig = {
	...baseConfig,
	mode: 'production',
	output: {
		path: path.resolve( __dirname, '../assets/js' ),
		chunkFilename: ( chunkData ) => getChunkName( chunkData, 'production' ),
		filename: '[name].js',
		// Prevents the collision of chunk names between different bundles.
		uniqueName: 'elementor',
	},
	performance: { hints: false },
};

const webpackProductionConfig = [
	{
		...prodSharedConfig,
		module: moduleRules,
		plugins: [
			...plugins,
		],
		name: 'base',
		entry: {
			// Clone.
			...entry,
		},
		optimization: {
			...prodSharedOptimization,
		},
	},
	{
		...prodSharedConfig,
		module: frontendModuleRules,
		plugins: [
			new RemoveChunksPlugin( '.bundle.min.js' ),
			...plugins,
		],
		name: 'frontend',
		entry: {
			// Clone.
			...frontendEntries,
		},
		optimization: {
			...prodSharedOptimization,
			runtimeChunk: {
				name: 'webpack.runtime.min',
			},
			splitChunks: {
				minChunks: 2,
			},
		},
	},
];

// Adding .min suffix to production entries.
webpackProductionConfig.forEach( ( config, index ) => {
	for ( const entryPoint in config.entry ) {
		let entryValue = config.entry[ entryPoint ];

		if ( entryValue.dependOn ) {
			// We duplicate the 'entryValue' obj for not affecting the 'entry' obj used by the dev process.
			entryValue = { ...entryValue };

			entryValue.dependOn += '.min';
		}

		delete config.entry[ entryPoint ];
		config.entry[ entryPoint + '.min' ] = entryValue;
	}
} );

// The 'packages' config doesn't need a `.min` suffix (it has its own config).
webpackProductionConfig.push( packagesConfigs.prod );

const developmentNoWatchConfig = webpackConfig.map( ( config ) => {
	return { ...config, watch: false };
} );

const productionWatchConfig = webpackProductionConfig.map( ( config ) => {
	return { ...config, watch: true };
} );

const gruntWebpackConfig = {
	development: webpackConfig,
	developmentNoWatch: developmentNoWatchConfig,
	production: webpackProductionConfig,
	productionWatch: productionWatchConfig,
	packages: packagesConfigs.dev,
};

module.exports = gruntWebpackConfig;
