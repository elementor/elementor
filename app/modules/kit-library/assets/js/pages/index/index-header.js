import Header from '../../components/layout/header';
import { ModalProvider, Heading, Text, Button } from '@elementor/app-ui';
import { useMemo, useState, useRef } from 'react';
import { useNavigate } from '@reach/router';
import PopoverDialog from 'elementor-app/ui/popover-dialog/popover-dialog';
import { appsEventTrackingDispatch } from 'elementor-app/event-track/apps-event-tracking';

import './index-header.scss';

export default function IndexHeader( props ) {
	const navigate = useNavigate();
	const [ isInfoModalOpen, setIsInfoModalOpen ] = useState( false );
	const importRef = useRef();
	const eventTracking = ( command, element = null, eventType = 'click', modalType = null ) => {
		appsEventTrackingDispatch(
			command,
			{
				element,
				event_type: eventType,
				page_source: 'home page',
				element_position: 'app_header',
				modal_type: modalType,
			},
		);
	};
	const onClose = ( e ) => {
		const element = e.target.classList.contains( 'eps-modal__overlay' ) ? 'overlay' : 'x';
		eventTracking( 'kit-library/modal-close', element, null, 'info' );
	};
	const buttons = useMemo( () => [
		{
			id: 'info',
			text: __( 'Info', 'elementor-pro' ),
			hideText: true,
			icon: 'eicon-info-circle-o',
			onClick: () => {
				eventTracking( 'kit-library/seek-more-info' );
				setIsInfoModalOpen( true );
			},
		},
		{
			id: 'refetch',
			text: __( 'Refetch', 'elementor-pro' ),
			hideText: true,
			icon: `eicon-sync ${ props.isFetching ? 'eicon-animation-spin' : '' }`,
			onClick: () => {
				eventTracking( 'kit-library/refetch' );
				props.refetch();
			},
		},
		{
			id: 'import',
			text: __( 'Import', 'elementor-pro' ),
			hideText: true,
			icon: 'eicon-upload-circle-o',
			elRef: importRef,
			onClick: () => {
				eventTracking( 'kit-library/kit-import' );
				navigate( '/import?referrer=kit-library' );
			},
		},
	], [ props.isFetching, props.refetch ] );

	return (
		<>
			<Header buttons={ buttons } />
			<PopoverDialog
				targetRef={ importRef }
				wrapperClass="e-kit-library__tooltip"
			>
				{ __( 'Import Kit', 'elementor' ) }
			</PopoverDialog>
			<ModalProvider title={ __( 'Welcome to the Library', 'elementor' ) }
				show={ isInfoModalOpen }
				setShow={ setIsInfoModalOpen }
				onOpen={ () => eventTracking( 'kit-library/modal-open', null, 'load', 'info' ) }
				onClose={ ( e ) => onClose( e ) }
			>
				<div className="e-kit-library-header-info-modal-container">
					<Heading tag="h3" variant="h3">{ __( 'What\'s a Website Kit?', 'elementor' ) }</Heading>
					<Text>{ __( 'A Website Kit is full, ready-made design that you can apply to your site. It includes all the pages, parts, settings and content that you\'d expect in a fully functional website.', 'elementor' ) }</Text>
				</div>
				<div className="e-kit-library-header-info-modal-container">
					<Heading tag="h3" variant="h3">{ __( 'What\'s going on in the Kit Library?', 'elementor' ) }</Heading>
					<Text>
						{ __( 'Search & filter for kits by category and tags, or browse through individual kits to see what\'s inside.', 'elementor' ) }
						<br />
						{ __( 'Once you\'ve picked a winner, apply it to your site!', 'elementor' ) }
					</Text>
				</div>
				<div>
					<Heading tag="h3" variant="h3">{ __( 'Happy browsing!', 'elementor' ) }</Heading>
					<Text>
						<Button
							url="https://go.elementor.com/app-kit-library-how-to-use-kits/"
							target="_blank"
							rel="noreferrer"
							text={ __( 'Learn more', 'elementor' ) }
							color="link"
							onClick={ () => {
								eventTracking( 'kit-library/seek-more-info', 'text link', null, 'info' );
							} }
						/>{ ' ' }
						{ __( 'about using templates', 'elementor' ) }
					</Text>
				</div>
			</ModalProvider>
		</>
	);
}

IndexHeader.propTypes = {
	refetch: PropTypes.func.isRequired,
	isFetching: PropTypes.bool,
};
