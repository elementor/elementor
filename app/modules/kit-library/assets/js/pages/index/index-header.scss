.e-kit-library {
	#eps-app-header-btn-refetch {
		padding: 0;
	}
}

.e-kit-library-header-info-modal-container {
	margin-bottom: spacing( 44 );
}

.e-kit-library__tooltip {
	padding: 5px 12px;
	color: $white;
	background-color: $color-gray-anthracite;
	font-size: 10px;

	&:before {
		border-bottom-color: $color-gray-anthracite;
	}
}
