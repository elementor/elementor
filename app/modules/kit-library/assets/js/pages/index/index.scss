$e-kit-library-sticky-search-z-index: 2;

.e-kit-library {
	&__index-layout-container {
		flex-grow: 1;
		overflow-y: auto;
		display: flex;
		flex-direction: column;
	}

	&__index-layout-top-area {
		padding: spacing(30) spacing(44);
		position: sticky;
		top: -1px;
		width: 100%;
		z-index: $e-kit-library-sticky-search-z-index;
		background-color: var(--app-background-color);
		gap: spacing( 24 );

		&-search, &-sort {
			min-width: 265px;
		}

		&-search {
			flex: 1;
		}
	}

	&__index-layout-main {
		padding-top: spacing(0);
		padding-bottom: spacing(24);
		overflow-y: hidden;
		height: auto;
		flex: 1 0 auto;
		display: flex;
		flex-direction: column;
		justify-content: space-between;
	}
}
