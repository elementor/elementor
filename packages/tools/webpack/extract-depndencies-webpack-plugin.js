// Inspired by "Dependency Extraction Webpack Plugin" by @wordpress team.
// Link: https://github.com/WordPress/gutenberg/tree/trunk/packages/dependency-extraction-webpack-plugin

const { sources: { RawSource } } = require( 'webpack' );

module.exports = class ExtractDependenciesWebpackPlugin {
	generateHandleName;
	replaceDependencyNames;
	generateAssetsFileName;

	constructor( {
		generateHandleName,
		replaceDependencyNames,
		generateAssetsFileName,
	} = {} ) {
		this.generateHandleName = generateHandleName || this.defaultGenerateHandleName;
		this.replaceDependencyNames = replaceDependencyNames || this.defaultReplaceDependencyNames;
		this.generateAssetsFileName = generateAssetsFileName || this.defaultGenerateAssetsFileName;
	}

	apply( compiler ) {
		compiler.hooks.thisCompilation.tap( this.constructor.name, ( compilation ) => {
			compilation.hooks.processAssets.tap( { name: this.constructor.name }, () => {
				[ ...compilation.entrypoints ].forEach( ( [ id, entrypoint ] ) => {
					const chunk = entrypoint.chunks.find( ( { name } ) => name === id );
					const chunkJSFile = this.getFileFromChunk( chunk );

					if ( ! chunkJSFile ) {
						return;
					}

					const deps = this.getDepsFromChunk( compilation, chunk );

					const assetFilename = this.generateAssetsFileName(
						compilation.getPath( '[file]', { filename: chunkJSFile } )
					);

					const content = this.createAssetsFileContent(
						id,
						deps,
						compilation.options.optimization.minimize
					);

					// Add source and file into compilation for webpack to output.
					compilation.assets[ assetFilename ] = new RawSource( content );

					chunk.files.add( assetFilename );
				} );
			} );
		} );
	}

	getDepsFromChunk( compilation, chunk ) {
		const externals = Object.keys( compilation.options.externals );
		const depsSet = new Set();

		compilation.chunkGraph.getChunkModules( chunk ).forEach( ( module ) => {
			[ ...( module.modules || [] ), module ].forEach( ( subModule ) => {
				const isExternalDep = externals.includes( subModule.userRequest );

				if ( ! isExternalDep ) {
					return;
				}

				depsSet.add( subModule.userRequest );
			} );
		} );

		return depsSet;
	}

	createAssetsFileContent( entryId, deps, shouldMinify ) {
		const handleName = this.generateHandleName( entryId );

		const depsAsString = [ ...deps ]
			.map( ( dep ) => this.replaceDependencyNames( dep ) )
			.filter( ( dep ) => dep !== handleName )
			.sort()
			.map( ( dep ) => `'${ dep }',` )
			.join( '\n\t\t' );

		let content =
`<?php
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

return [
	'handle' => '${ handleName }',
	'deps' => [
		${ depsAsString }
	],
];
`;

		if ( shouldMinify ) {
			content = content
				.replaceAll( /\s/gm, '' )
				.replace( '<?php', '<?php\n' );
		}

		return content;
	}

	getFileFromChunk( chunk ) {
		return [ ...chunk.files ].find( ( f ) => /\.js$/i.test( f ) );
	}

	defaultReplaceDependencyNames( name ) {
		const map = new Map( [
			[ '@elementor/', 'elementor-packages-' ],
			[ '@wordpress/', 'wp-' ],
		] );

		for ( const [ key, value ] of map ) {
			if ( name.startsWith( key ) ) {
				return name.replace( key, value );
			}
		}

		return name;
	}

	defaultGenerateAssetsFileName( filename ) {
		return filename.replace( /(\.min)?\.js$/i, '.asset.php' );
	}

	defaultGenerateHandleName( name ) {
		return `elementor-packages-${ name }`;
	}
};
