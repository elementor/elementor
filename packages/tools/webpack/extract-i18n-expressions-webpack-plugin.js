const fs = require( 'fs' );
const path = require( 'path' );

const MODULE_FILTERS = Object.freeze( [ /(([^!?\s]+?)(?:\.js|\.jsx|\.ts|\.tsx))$/, /^((?!node_modules).)*$/ ] );

const COMMENTS_REGEXPS = Object.freeze( [
	// Matches translators comment block: `/* translators: %s */`.
	/\/\*[\t ]*translators:.*\*\//gm,
	/// Matches translators inline comment: `// translators: %s`.
	/(\/\/)[\t ]*translators:[^\r\n]*/gm,
] );

const TRANSLATIONS_REGEXPS = Object.freeze( [
	// Matches translation functions: `__('Hello', 'elementor')`, `_n('Me', 'Us', 2, 'elementor-pro')`.
	/\b_(?:_|n|nx|x)\(.*?,\s*(?<c>['"`])[\w-]+\k<c>\)/gm,
] );

module.exports = class ExtractI18nExpressionsWebpackPlugin {
	apply( compiler ) {
		// Learn more about Webpack plugin system: https://webpack.js.org/api/plugins/

		let translationCallExpressions = [];

		// Learn more about Webpack compilation process and hooks: https://webpack.js.org/api/compilation-hooks/
		compiler.hooks.thisCompilation.tap( this.constructor.name, ( compilation ) => {
			// We tap into the time that Webpack has finished processing all the other assets
			// learn more: https://webpack.js.org/api/compilation-hooks/#processassets.
			compilation.hooks.processAssets.tap( { name: this.constructor.name }, () => {
				translationCallExpressions = this.getModuleExpressionsMap( compilation );
			} );
		} );

		compiler.hooks.afterEmit.tapPromise( this.constructor.name, async ( compilation ) => {
			// Create all the translations files based on the call expressions.
			await this.createTranslationsFiles( compilation, translationCallExpressions );
		} );
	}

	getModuleExpressionsMap( compilation ) {
		const moduleExpressionsMap = new Map();

		[ ...compilation.chunks ].forEach( ( chunk ) => {
			const chunkJSFile = this.getFileFromChunk( chunk );

			if ( ! chunkJSFile ) {
				// There's no JS file in this chunk, no work for us. Typically a `style.css` from cache group.
				return;
			}

			compilation.chunkGraph.getChunkModules( chunk ).forEach( ( module ) => {
				this.getSubModulesToCheck( module ).forEach( ( subModule ) => {
					const mainEntryFile = this.findMainModuleOfEntry( subModule, compilation );

					if ( ! moduleExpressionsMap.has( mainEntryFile ) ) {
						moduleExpressionsMap.set( mainEntryFile, [] );
					}

					// Running over the submodules and find all the translation call expressions and their translators comment
					// (e.g `/* translators: %s: name*/ __('Hello %s', 'elementor')`),
					// extract them and add them to a Map, where the key is the main entry file, and the value is an array of all the
					// translation call expressions.
					this.extractExpressionsFromSubmodule( subModule ).forEach( ( expression ) => {
						moduleExpressionsMap.get( mainEntryFile ).push( expression );
					} );
				} );
			} );
		} );

		return moduleExpressionsMap;
	}

	async createTranslationsFiles( compilation, translationCallExpressions ) {
		const promises = [ ...compilation.entrypoints ].map( ( [ id, entrypoint ] ) => {
			const chunk = entrypoint.chunks.find( ( { name } ) => name === id );

			const chunkJSFile = this.getFileFromChunk( chunk );

			if ( ! chunkJSFile ) {
				return Promise.resolve();
			}

			const mainFilePath = compilation.options.entry[ id ].import[ 0 ];

			const assetFilename = this.generateTranslationFilename(
				compilation.options.output.path,
				compilation.getPath( '[file]', { filename: chunkJSFile } )
			);

			const assetFileContent = this.generateTranslationFileContent(
				translationCallExpressions.get( mainFilePath ) || []
			);

			return fs.promises.writeFile( assetFilename, assetFileContent );
		} );

		return await Promise.all( promises );
	}

	getSubModulesToCheck( module ) {
		return [ ...( module.modules || [] ), module ]
			.filter( ( subModule ) => this.shouldCheckModule( subModule ) );
	}

	getFileFromChunk( chunk ) {
		return [ ...chunk.files ].find( ( f ) => /\.js$/i.test( f ) );
	}

	shouldCheckModule( module ) {
		return MODULE_FILTERS.every( ( filter ) => !! filter.test( module.userRequest ) );
	}

	findMainModuleOfEntry( module, compilation ) {
		if ( compilation.moduleGraph.getIssuer( module ) ) {
			return this.findMainModuleOfEntry( compilation.moduleGraph.getIssuer( module ), compilation );
		}

		return module.rawRequest;
	}

	extractExpressionsFromSubmodule( subModule ) {
		const source = subModule?._source?._valueAsString;

		if ( ! source ) {
			return [];
		}

		const translationCallExpressions = [];

		[
			...TRANSLATIONS_REGEXPS,
			...COMMENTS_REGEXPS,
		].forEach( ( regexp ) => {
			[ ...source.matchAll( regexp ) ].forEach( ( res ) => {
				translationCallExpressions.push( {
					type: COMMENTS_REGEXPS.includes( regexp ) ? 'comment' : 'call-expression',
					index: res.index,
					value: res[ 0 ],
				} );
			} );
		} );

		return translationCallExpressions;
	}

	generateTranslationFilename( basePath, filename ) {
		return path.join(
			basePath,
			filename.replace( /(\.min)?\.js$/i, '.strings.js' )
		);
	}

	generateTranslationFileContent( expressions ) {
		return expressions
			// Sort by the index it was found in the file based on the regexp (and not by the order it was added to the array).
			.sort( ( a, b ) => a.index - b.index )
			// Add a semicolon when needed.
			.map( ( expr ) => `${ expr.value }${ expr.type === 'comment' ? '' : ';' }` )
			// Join all the expressions to a single string with line-breaks between them.
			.join( '\n' );
	}
};
