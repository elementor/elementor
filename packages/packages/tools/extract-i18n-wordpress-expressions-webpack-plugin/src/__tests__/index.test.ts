/**
 * @jest-environment node
 */

import * as fs from 'fs';
import { vol } from 'memfs';
import * as path from 'path';
import { webpack } from 'webpack';

import { ExtractI18nWordpressExpressionsWebpackPlugin } from '../index';

jest.mock( 'fs', () => jest.requireActual( 'memfs' ) );

describe( '@elementor/extract-i18n-wordpress-expressions-webpack-plugin', () => {
	beforeEach( () => {
		// Entry file.
		fs.mkdirSync( path.resolve( '/app/dist' ), { recursive: true } );
		fs.writeFileSync( path.resolve( '/app/dist/index.js' ), '' );

		fs.mkdirSync( path.resolve( '/app2/dist' ), { recursive: true } );
		fs.writeFileSync( path.resolve( '/app2/dist/index.js' ), '' );

		// Should be in output.
		fs.mkdirSync( path.resolve( '/app/src' ), { recursive: true } );
		fs.mkdirSync( path.resolve( '/app/src/components' ), { recursive: true } );
		fs.mkdirSync( path.resolve( '/app/src/hooks' ), { recursive: true } );

		fs.writeFileSync(
			path.resolve( '/app/src/components/component.js' ),
			`
			export default function Component() {
				return __( "Component", 'elementor' );
			}
		`
		);

		fs.writeFileSync(
			path.resolve( '/app/src/hooks/hook.ts' ),
			`
			export default function useHook() {
				return __('hook','elementor');
			}
		`
		);

		fs.writeFileSync(
			path.resolve( '/app/src/index.js' ),
			`
			export default function Index() {
				__(
					'Some long text with multiple lines in the function call',
					'elementor'
				);

				__( "Unique domain", "some-plugin-slug" );

				// translators: %1$s - special placeholder.
				const withSpecialPlaceHolder = __('special placeholder %1$s','elementor' );

				// translators: %s - regular comment.
				const withComment = __( 'regular comment %s','elementor');

				/* translators: %s - comment block. */
				const withCommentBlock = __('comment block %s', 'elementor');

				console.log(__( 'inside console log', 'elementor' ))

				return [
					_n( 'basic with', 'plural', 2, 'elementor' ),
					_nx( 'another with' , 'plural' , 2 , 'elementor' ),
					_x( 'context', 'elementor' ),
					invalid__( 'invalid', 'elementor' ),
				];
			}
		`
		);

		fs.mkdirSync( path.resolve( '/app2/src' ), { recursive: true } );
		fs.writeFileSync(
			path.resolve( '/app2/src/index.js' ),
			`
			export const test = __( 'from app 2', 'elementor' );
		`
		);

		// Should not be in output.
		const ignoredContent = `
			export default function ShouldIgnoreComponent() {
				return __('should ignore','elementor');
			}
		`;

		fs.mkdirSync( path.resolve( '/app/fake-src' ), { recursive: true } );
		fs.writeFileSync( path.resolve( '/app/fake-src/index.js' ), ignoredContent );

		fs.mkdirSync( path.resolve( '/app/src/__tests__' ), { recursive: true } );
		fs.writeFileSync( path.resolve( '/app/src/__tests__/mock.test.js' ), ignoredContent );

		fs.writeFileSync( path.resolve( '/app/src/not-a-js-file.txt' ), ignoredContent );
	} );

	afterEach( () => {
		vol.reset();
	} );

	it( 'should extract translations from scripts', ( done ) => {
		// Arrange.
		const compiler = webpack( {
			mode: 'development',
			entry: {
				app: path.resolve( '/app/dist/index.js' ),
				app2: path.resolve( '/app2/dist/index.js' ),
			},
			output: {
				filename: '[name]/[name].js',
				path: path.resolve( '/output' ),
			},
			plugins: [
				new ExtractI18nWordpressExpressionsWebpackPlugin( {
					pattern: ( entryPath ) => path.resolve( entryPath, '../../src/**/*.{ts,tsx,js,jsx}' ),
				} ),
			],
		} );

		// Act.
		compiler.run( ( err, stats ) => {
			// Assert.
			expect( err ).toBe( null );
			expect( stats?.hasErrors() ).toBe( false );
			expect( stats?.hasWarnings() ).toBe( false );

			expect(
				fs.readFileSync( path.resolve( `/output/app/app.strings.js` ), { encoding: 'utf8' } )
			).toMatchSnapshot();

			expect(
				fs.readFileSync( path.resolve( `/output/app2/app2.strings.js` ), { encoding: 'utf8' } )
			).toMatchSnapshot();

			done();
		} );
	} );
} );
