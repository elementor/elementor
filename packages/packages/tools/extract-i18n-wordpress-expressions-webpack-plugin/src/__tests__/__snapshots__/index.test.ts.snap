// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@elementor/extract-i18n-wordpress-expressions-webpack-plugin should extract translations from scripts 1`] = `
"__(
					'Some long text with multiple lines in the function call',
					'elementor'
				);
__( "Unique domain", "some-plugin-slug" );
// translators: %1$s - special placeholder.
__('special placeholder %1$s','elementor' );
// translators: %s - regular comment.
__( 'regular comment %s','elementor');
/* translators: %s - comment block. */
__('comment block %s', 'elementor');
__( 'inside console log', 'elementor' );
_n( 'basic with', 'plural', 2, 'elementor' );
_nx( 'another with' , 'plural' , 2 , 'elementor' );
_x( 'context', 'elementor' );
__('hook','elementor');
__( "Component", 'elementor' );"
`;

exports[`@elementor/extract-i18n-wordpress-expressions-webpack-plugin should extract translations from scripts 2`] = `"__( 'from app 2', 'elementor' );"`;
